File: 05_Development/CareBridgeMobileApp/lib/features/directChat/calls/direct_call_state.dart
import '../models/conversation_call.dart';

enum DirectCallPhase {
  idle,
  outgoing,
  incoming,
  answering,
  readyToJoin,
  joining,
  inCall,
  reconnecting,
  terminal,
  failed,
}

class DirectCallState {
  final DirectCallPhase phase;
  final ConversationCall? call;
  final String? message;

  const DirectCallState({
    this.phase = DirectCallPhase.idle,
    this.call,
    this.message,
  });

  bool get isVideo => call?.callType == 'VIDEO';
}

const _terminalStatuses = {
  'DECLINED',
  'MISSED',
  'CANCELLED',
  'ENDED',
  'FAILED',
};

bool shouldCancelOutgoingOnDetach({
  required ConversationCall? call,
  required String? currentUserId,
}) {
  return call != null &&
      currentUserId != null &&
      call.initiatedByUserId == currentUserId &&
      (call.callStatus == 'INITIATED' || call.callStatus == 'RINGING');
}

DirectCallState reduceCallState({
  required String currentUserId,
  required ConversationCall call,
  DirectCallState? previous,
}) {
  if (previous?.call?.callId == call.callId &&
      _terminalStatuses.contains(previous!.call!.callStatus) &&
      !_terminalStatuses.contains(call.callStatus)) {
    return previous;
  }
  if (_terminalStatuses.contains(call.callStatus)) {
    return DirectCallState(phase: DirectCallPhase.terminal, call: call);
  }
  if (call.callStatus == 'ANSWERED') {
    if (previous?.call?.callId == call.callId &&
        {
          DirectCallPhase.joining,
          DirectCallPhase.inCall,
          DirectCallPhase.reconnecting,
        }.contains(previous!.phase)) {
      return DirectCallState(
        phase: previous.phase,
        call: call,
        message: previous.message,
      );
    }
    return DirectCallState(phase: DirectCallPhase.readyToJoin, call: call);
  }
  final isCaller = call.initiatedByUserId == currentUserId;
  return DirectCallState(
    phase: isCaller ? DirectCallPhase.outgoing : DirectCallPhase.incoming,
    call: call,
  );
}

File: 05_Development/CareBridgeWebApp/src/features/directChat/calls/directCallState.ts
import type { ConversationCall } from '../models/directConversation';

export type DirectCallPhase =
  | 'idle'
  | 'outgoing'
  | 'incoming'
  | 'answering'
  | 'readyToJoin'
  | 'joining'
  | 'inCall'
  | 'reconnecting'
  | 'terminal'
  | 'failed';

export interface DirectCallState {
  phase: DirectCallPhase;
  call: ConversationCall | null;
  message?: string;
}

export const idleDirectCallState: DirectCallState = {
  phase: 'idle',
  call: null,
};

const terminalStatuses = new Set([
  'DECLINED',
  'MISSED',
  'CANCELLED',
  'ENDED',
  'FAILED',
]);

export function reduceDirectCallState(
  currentUserId: string,
  call: ConversationCall,
  previous: DirectCallState = idleDirectCallState
): DirectCallState {
  if (
    previous.call?.callId === call.callId &&
    terminalStatuses.has(previous.call.callStatus) &&
    !terminalStatuses.has(call.callStatus)
  ) {
    return previous;
  }
  if (terminalStatuses.has(call.callStatus)) {
    return { phase: 'terminal', call };
  }
  if (call.callStatus === 'ANSWERED') {
    if (
      previous.call?.callId === call.callId &&
      ['joining', 'inCall', 'reconnecting'].includes(previous.phase)
    ) {
      return { ...previous, call };
    }
    return { phase: 'readyToJoin', call };
  }
  return {
    phase: call.initiatedByUserId === currentUserId ? 'outgoing' : 'incoming',
    call,
  };
}
